include/simdjson/padded_string.h

```diff
@@ -33,77 +50,160 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
 // constructors.
 struct padded_string final {
 
+    using type = padded_string;
+    using value_type = char *;
+  // free the previous payload
+  // allocate for the new size
+  // return false on ENOMEM
+  static bool reset(type & pstring_, size_t new_size_) noexcept 
+  {
+      if (pstring_.data())
+        pstring_.~type(); 
+
+    pstring_.viable_size = new_size_;
+      pstring_.data_ptr = allocate_padded_buffer(new_size_);
+    if (pstring_.data_ptr != nullptr) {
+        pstring_.data_ptr[new_size_] = '\0'; // easier when you need a c_str
+      return true;
+    }
+    // failed to allocate
+    pstring_.viable_size = 0;
+    return false;
+  }
+
   explicit padded_string() noexcept : viable_size(0), data_ptr(nullptr) {}
 
-  explicit padded_string(size_t length) noexcept
-      : viable_size(length), data_ptr(allocate_padded_buffer(length)) {
-    if (data_ptr != nullptr)
-      data_ptr[length] = '\0'; // easier when you need a c_str
+  explicit padded_string(size_t length) noexcept 
+      : data_ptr(allocate_padded_buffer(length)), viable_size(length)
+  {
+    #ifndef NDEBUG
+    if (data_ptr != nullptr) {
+      /* do what? */
+    }
+    #endif
   }
 
-  explicit padded_string(char *data, size_t length) noexcept
-      : viable_size(length), data_ptr(allocate_padded_buffer(length)) {
+  explicit padded_string(const char *data, size_t length) noexcept
+      : padded_string(length) 
+  {
     if ((data != nullptr) and (data_ptr != nullptr)) {
       memcpy(data_ptr, data, length);
-      data_ptr[length] = '\0'; // easier when you need a c_str
+      // probably not necessary
+      data_ptr[length] = '\0'; 
+    } else {
+      /* do what? */
```

Comment:
The C++ thing to do would be to throw an exception. In our case, I recommend simply assuming that data_ptr might be nullptr and unallocated. I suggest setting the viable_length to zero.

Reply:
OK